Mako Batch/Interactive Templates & Build Fixes

henryleberre opened the pull request:
The last PR before the Benchmarking one.

Closes #292, #240, and #287.

sbryngelson commented on docs/documentation/running.md:
```diff
@@ -2,7 +2,31 @@
+>
+> Adding a new template file or modifying an existing one will most likely be required if:
+> - You are on a cluster that does not have a template yet.
+> - Your cluster is configured with SLURM and but fails when interactive jobs are
```
typo 'and but'

henryleberre replied:
Fixed! I am getting tired..

sbryngelson commented on misc/run-phoenix-release-cpu.sh:
```diff
@@ -12,5 +12,4 @@ cd "$SLURM_SUBMIT_DIR"
 . ./mfc.sh load -c p -m gpu
-./mfc.sh test -j $(nproc) -b mpirun -a
-
+./mfc.sh test -j $(nproc) -a -- -b mpirun
```
can you comment on what `--` does?

henryleberre replied:
It forwards arguments to `./mfc.sh run` (called by `./mfc.sh test`). This means we no longer have to redocument all of these options as options of `./mfc.sh test`, too. We also no longer have to manually forward each option.

henryleberre replied:
I just changed this line to:
```console
./mfc.sh test -j $(nproc) -a -- -c phoenix
```

sbryngelson commented on toolchain/mfc/args.py:
```diff
@@ -98,29 +94,31 @@ def add_common_arguments(p, mask = None):
-    run.add_argument("--wait",                 action="store_true",                       default=False,      help="(Batch) Wait for the job to finish.")
+    run.add_argument("input",                      metavar="INPUT",              type=str,                     help="Input file to run.")
+    run.add_argument("arguments",                  metavar="ARGUMENTS", nargs="*", type=str, default=[],       help="Additional positional arguments to pass to the case file.")
+    run.add_argument("-e", "--engine",             choices=["interactive", "batch"],              type=str, default="interactive", help="Job execution/submission engine choice.")
```
is there an option to use `-b mpirun` for template files or does one have to modify the template

henryleberre replied:
I agree, I just added it back in.

sbryngelson commented on toolchain/mfc/run/run.py:
```diff
@@ -24,30 +31,67 @@ def validate_job_options() -> None:
 
-    engine = engines.get_engine(ARG("engine"))
-    engine.init(input_file)
+        return ["nsys", "profile", "--stats=true", "--trace=mpi,nvtx,openacc"] + ARG("nsys")
```
i think i had a comment on your original PR that said that nsys profile needs to go in the opposite order against `mpirun` for whatever we do currently (for mpi jobs at least). does this fix that?

henryleberre replied:
Noted. I just switched the order.

sbryngelson commented on toolchain/templates/default.mako:
```diff
@@ -0,0 +1,53 @@
+    % if not mpi:
+        ${' '.join([f"'{x}'" for x in profiler ])} "${binpath}"
+    % else:
+        if command -v jsrun > /dev/null; then
```
can we not just pass `-b` in here? i know we should be using the templates but this seems like an easy improvement of sorts.

henryleberre replied:
As per above, I agree and I just added it.

sbryngelson commented:
I think this PR was going to fix the issue of not using multiple ranks with `./mfc.sh test -j X` on the Phoenix CPU runner. does it?

henryleberre commented:
> I think this PR was going to fix the issue of not using multiple ranks with `./mfc.sh test -j X` on the Phoenix CPU runner. does it?

I had forgotten to include it. Just did. I added:
```console
               --bind-to none                             \
```

So when using GT Phoenix, you just need to use `-c phoenix` while testing and/or running (interactive & batch) MFC.

sbryngelson commented:
so the new command to test on an `mpirun` system is `./mfc.sh test -j X -- --binary mpirun --gpu`? this seems to be what I need when testing your PR on Phoenix.

henryleberre commented:
> so the new command to test on an `mpirun` system is `./mfc.sh test -j X -- --binary mpirun --gpu`? this seems to be what I need when testing your PR on Phoenix.
> 
> it is also re-building even though the GLOBs are the same?

For Phoenix, I would recommend just:
```console
./mfc.sh test -j X --gpu -- -c phoenix
```

I looked into #288 and it's not quite fixed yet. It is really quite strange. I think it has to do with CMake expecting its binary directory to be empty upon creation skipping GLOB checks within that directory.

sbryngelson commented:
> > so the new command to test on an `mpirun` system is `./mfc.sh test -j X -- --binary mpirun --gpu`? this seems to be what I need when testing your PR on Phoenix.
> > it is also re-building even though the GLOBs are the same?
> 
> For Phoenix, I would recommend just:
> 
> ```
> ./mfc.sh test -j X --gpu -- -c phoenix
> ```
> 
> But this also works now:
> 
> ```
> ./mfc.sh test -j X --gpu -- -b mpirun -f --bind-to none
> ```
> 
> I looked into #288 and it's not quite fixed yet. It is really quite strange. I think it has to do with CMake expecting its binary directory to be empty upon creation and thus skipping GLOB checks within that directory.

got it

henryleberre commented:
Part of the appeal of this PR (other than there being a template per system) is that:
- Templates are now used for both Batch & Interactive execution. This gives us more flexibility and makes it a lot easier to understand and modify the behavior of these template files. You no longer need to parse through `toolchain/mfc/run/engines.py`. It doesn't exist anymore!
- Instead of my handcrafted and regex-powered template engine, I switched to an all-around better solution, Mako.
- A lot of indirection was removed (through the support of `--` in argument parsing in many places).
- Refactored `system()` in `toolchain/mfc/` so that whenever we execute a command, we no longer depend on the shell. We provide a list of arguments rather a command-string. It's a lot more difficult to create a bug by introducing spaces in a command or variable name.

sbryngelson commented:
This looks great - glad to merge once tests pass. We can also now run CPU tests much faster.

henryleberre commented:
@sbryngelson I just fixed the rebuild issue! Closes #288.

sbryngelson commented:
> @sbryngelson I just fixed the rebuild issue! Closes #288.

Thank goodness! It was very strange. What was it?

sbryngelson commented:
@henryleberre you have access to NCSA Delta and Purdue Anvil as well. Can you add appropriate batch templates for those computers?

henryleberre commented:
> > @sbryngelson I just fixed the rebuild issue! Closes #288.
> 
> Thank goodness! It was very strange. What was it?

There were two issues:
- We only generated the case-optimization FPP files while building, not also while configuring. This causes a GLOB mismatch.
- When I added their generation to the configure step, I had forgotten we also delete the binary directory before running CMake (we have good reasons for doing this). So the files were actually never there while CMake was configuring the MFC targets.

henryleberre commented:
> @henryleberre you have access to NCSA Delta and Purdue Anvil as well. Can you add appropriate batch templates for those computers?

Could these be added later? It might be a good idea for users to get acquainted with this way of doing things, adding Mako files as they need them.

sbryngelson commented:
> > @henryleberre you have access to NCSA Delta and Purdue Anvil as well. Can you add appropriate batch templates for those computers?
> 
> Could these be added later? It might be a good idea for users to get acquainted with this way of doing things, adding Mako files as they need them.

Yes that's fine. Merge when appropriate.
